File: Replication-code-(Python)/dupire.py
import numpy as np 

def dupire(prices, t, k, K):
    """ 
    Computing the local volatility grid for the Dupire Volatility 
    Adjustments in the end to extrapolate over the areas that couldn't be computed
    """
    
    dC_dT = np.gradient(prices, t, axis=0)  # Time derivative
    dC_dT = np.where(dC_dT < 0, 1e-10, dC_dT) # can't be 0 given Dupire formula

    
    dC_dK = np.gradient(prices, k, axis=1)  # First strike derivative
    d2C_dK2 = np.gradient(dC_dK, k, axis=1) # Second strike derivative

    local_vol = np.sqrt((dC_dT) / (0.5 * K**2 * d2C_dK2))

    return local_vol

File: Replication-code/functions/dupire.m
function local_vol = dupire(prices, par)
    % Computing the local volatility grid for the Dupire Volatility 
    % Adjustments in the end to extrapolate over the areas that couldn't be computed

    grad = gradient(prices, par.tgrid_train, par.kgrid_train, 2);
    dC_dT = grad{1};
    dC_dK = grad{2};
    dC_dT(dC_dT < 0) = 1e-10;

    grad_2 = gradient(dC_dK, par.tgrid_train, par.kgrid_train, 2);
    d2C_dK2 = grad_2{2};

    local_vol = sqrt((dC_dT) ./ (0.5 * par.K_train.^2 .* d2C_dK2));
    local_vol(d2C_dK2 < 0) = NaN;
    local_vol(local_vol == Inf) = NaN;
    % local_vol = min(local_vol, 1);

end


